flink-streaming-java/src/main/java/org/apache/flink/streaming/api/environment/ExecutionCheckpointingOptions.java

```diff
@@ -314,4 +317,28 @@ public class ExecutionCheckpointingOptions {
                             "Defines the maximum number of subtasks that share the same channel state file. "
                                     + "It can reduce the number of small files when enable unaligned checkpoint. "
                                     + "Each subtask will create a new channel state file when this is configured to 1.");
+
+    public static final ConfigOption<ThreadDumpLogLevel>
+            THREAD_DUMP_LOG_LEVEL_WHEN_CHECKPOINT_TIMEOUT =
+                    key("execution.checkpointing.thread-dump-log-level-when-checkpoint-timeout")
```

Comment:
I think printing the thread dump info is not only useful when checkpoint expires, we can also enable it if job is in high back-pressured. Thus, I suggest to split this parameter into two parameters. One is for `thread-dump-log-level`, and other is for the scenarios. Such as we can print the thread dump info when checkpoint expires, task is running so busy, task is failing or the process will exit for fatal error. 
We can only support this feature for expired checkpoint in this PR.

Reply:
That's a very good point. And is it meaningful to define different log level for different scenarios? Thus I prefer change the option name to `execution.checkpointing.thread-dump-log-level.checkpoint-expired`, and provide different suffixes for various scenarios. WDYT?

Reply:
First of all, I think not all scenarios obey the `execution.checkpointing` prefix. And I think making different log level for different scenarios would be a bit too complex. We only need two parameters no matter how many scenarios supported in my proposal.